### PR TITLE
Refactor: use builder pattern with socket options

### PR DESCRIPTION
**Pull request by jonasbostoen**
This PR implements the builder pattern for all the socket options and uses that in the examples & benchmarks. It also fixed a bug where the `max_clients` option was not respected on server side sockets.

### CODE REVIEW

**Comment by quangkeu95** [comment]
Wondering why don't you use the lib [derive-builder](https://docs.rs/derive_builder/latest/derive_builder) to implement the builder pattern? Sorry if this is a stupid question, just wanna know what factors that you consider when do some refactors like this.

**Comment by jonasbostoen** [comment]
@quangkeu95 that was also definitely an option, although with derive-builder you're more constrained

**Comment by quangkeu95** [comment]
Make sense, so it's better to avoid using extra dependencies for building low-level lib.

File: msg-socket/src/req/mod.rs [hunk 1]
```diff
@@ -40,39 +40,77 @@ pub enum Command {
 
 #[derive(Debug, Clone)]
 pub struct ReqOptions {
-    pub auth_token: Option<Bytes>,
-    pub timeout: std::time::Duration,
-    pub retry_on_initial_failure: bool,
-    pub backoff_duration: std::time::Duration,
+    auth_token: Option<Bytes>,
+    timeout: std::time::Duration,
+    blocking_connect: bool,
```
Comment (merklefruit) [inline msg-socket/src/req/mod.rs:45]: can you add inline docs for this?

**Review by merklefruit — APPROVED** [review]
LGTM! nice

**Pull request merged**